### ISSUE TITLE
Create prop onSuccess has wrong typing

### ISSUE THREAD

**Issue by HendrikRoehm** [marmelab/react-admin#5761]
According to the [CreateProps typings](https://github.com/marmelab/react-admin/blob/872480a9a4a8491066eb8b1f0f651e48f47e9080/packages/ra-ui-materialui/src/types.ts#L50), the onSuccess callback has the type `(data: Record) => void`.

However, it seems that the true implementation in the [useCreateController](https://github.com/marmelab/react-admin/blob/872480a9a4a8491066eb8b1f0f651e48f47e9080/packages/ra-core/src/controller/details/useCreateController.ts#L167) the type `({ data } : { data: Record}) => void` is used.

Right now the two are incompatible. I would suggest to change the type as this breaks no implementation based on the current behaviour?

**Comment by fzaninotto** [comment]
Very well spotted! It is indeed a bug. Would you like to open a PR to patch the types?

**Comment by HendrikRoehm** [comment]
Sure, I will create a PR.

**Comment by HendrikRoehm** [comment]
I have opened a pull-request. During my investigation in the code, it seems that there a minimum of 3 different usages of an onSuccess property in different locations. Additionally, there were a lot of locations where basically no typing was present (everything was any). I tried to remove some abiguity by defining more types.

I was not quite sure how to test that, so I am happy for suggestions on how to improve the PR.
